added killing containers with other signals

MarcoLotz opened the pull request:
fixes #614
added Signal types for killing containers

johnflavin commented on src/main/java/com/spotify/docker/client/DockerClient.java:
```diff
@@ -1003,6 +1001,39 @@ void restartContainer(String containerId, int secondsToWaitBeforeRestart)
+  /**
+   * Supported parameters for {@link #killContainer(String, TerminationSignals)}).
+   */
+  enum TerminationSignals {
```
This should be named `TerminationSignal`, singular.

MarcoLotz commented:
@johnflavin changed.

mattnworb commented on src/main/java/com/spotify/docker/client/DefaultDockerClient.java:
```diff
@@ -715,10 +715,31 @@ public void restartContainer(String containerId, int secondsToWaitBeforeRestart)
+    checkNotNull(containerId, "containerId");
+
+    try {
+      final WebTarget resource = noTimeoutResource()
```
why use the `noTimeout` client here?

seems like there is also room to re-use containerAction, by refactoring it to have an overload where you can pass additional query params, to maintain the idea that most of these methods that call `POST /containers/:id/:action` flow through the same method

MarcoLotz replied:
Good point. Will take a look into it.

mattnworb commented on src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java:
```diff
@@ -1366,6 +1366,72 @@ public void testRestartContainer() throws Exception {
+
+    // Should not be running
+    {
+      final ContainerInfo containerInfoLatest = sut.inspectContainer(containerId);
```
just curious: why wrap this in a block if you use different names for the ContainerInfo instances and therefore there is no name collision?

mattnworb commented on src/main/java/com/spotify/docker/client/DockerClient.java:
```diff
@@ -990,9 +990,7 @@ void restartContainer(String containerId, int secondsToWaitBeforeRestart)
    * Kill a docker container.
-   * Note: This implementation deviates from the Docker Remote API. The
-   * latter accepts the kill signal as an argument. This implementation does
-   * not accept the signal argument; instead, the default SIGKILL is sent.
```
good improvement here, thanks

MarcoLotz commented on src/main/java/com/spotify/docker/client/DockerClient.java:
```diff
@@ -1003,6 +1001,39 @@ void restartContainer(String containerId, int secondsToWaitBeforeRestart)
+    SIGTERM("SIGTERM"),
+    SIGKILL("SIGKILL"),
+    SIGSTOP("SIGSTOP");
+
```
Well spotted. Totally agree with you. Any other signal? I can add all (http://www.comptechdoc.org/os/linux/programming/linux_pgsignals.html) just to be future friendly.

albers replied:
Here's what Dockers bash completion uses:
```bash
        local signals=(
                SIGCONT
                SIGHUP
                SIGINT
                SIGKILL
                SIGQUIT
                SIGSTOP
                SIGTERM
                SIGUSR1
                SIGUSR2
        )
```
This is quite an arbitrary subset of the available signals, but nobody ever complained so far.

albers replied:
I think it's best to add all signals because else there would be no way to specify a value missing in the enum. (for bash completion, it's different because the user can enter any value.)

If the enum offeres a subset of the available signals, there should be an overloaded `String` variant that allows passing the missing signals. I prefer the first solution.

MarcoLotz commented:
@albers: added all the signals and changed naming.

@mattnworb:
i) added overloaded definition to containerAction. Originally I had defined a nested class for the query parameters and used varargs in the containerAction signature. I've changed to Map.Entry to avoid private nested classes. Changed from varargs of Map.Entry<String, String> to a List of Map.Entry<String,String> because was getting many warnings on type inference. The only solution would have been to add a SuppressWarnings on it. Should you decide that going for the nested Query Parameters class and vargs is a better solution (or anything like it) let me know and I will update it.

ii) Updated the restartContainer method to use the new containerAction. Should any further refactoring be required in order to use the new containerAction method, probably would be better to do in another pull request

johnflavin commented:
Instead of a `List<Map.Entry<String, String>>`, couldn't you accomplish the same thing by just having the method take a `Map<String, String>`?

MarcoLotz commented:
Well noted. To use a map, we would have to assume that each query parameter field is unique. It may be valid now for docker API, but there's no spec forcing query parameters fields to be unique (http://stackoverflow.com/questions/1746507/authoritative-position-of-duplicate-http-get-query-keys). That's why I used a List instead. I totally agree that Map would be a lot cleaner.



Sent from my iPhone
> On 14 Apr 2017, at 11:12, Flavin <notifications@github.com> wrote:
> 
> Instead of a List<Map.Entry<String, String>>, couldn't you accomplish the same thing by just having the method take a Map<String, String>?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

johnflavin commented:
You’re correct that for general HTTP APIs, query params aren't guaranteed to have unique keys. But they are unique in docker's API (at least as it exists so far). When docker's API needs to allow multiple values for one parameter, as can be the case with `filter` params, they want the params in the form `name=[val1, val2,...]` rather than `name=val1&name=val2&...`.

mattnworb commented:
other places in the docker-client codebase use Jersey's `MultivaluedMap` for just this purpose

MarcoLotz commented:
@johnflavin @mattnworb awesome! This makes things a lot cleaner and simpler. Will do the changes.

davidxia commented:
@mattnworb Is this good to go? Do we want `Signal` to be an inner enum or in a separate file?

mattnworb commented:
@davidxia doesn't make much difference to me.
